File: src/encode.zig
const std = @import("std");
const constants = @import("constants.zig");
const fir = @import("fir.zig");
const subcarrier = @import("subcarrier.zig");

const uv_taps_pal = [_]i16{ 4, 38, 270, 1226, 3619, 6927, 8600, 6927, 3619, 1226, 270, 38, 4 };
const uv_taps_ntsc = [_]i16{ 69, 626, 2959, 7563, 10334, 7563, 2959, 626, 69 };

const kb = 0.49211104112248356308804691718185;
const kr = 0.87728321993817866838972487283129;

pub const Encoder = struct {
    pub const Error = error{UnsupportedStandard};

    standard: constants.Standard,
    width: usize,
    denominator: usize,
    precombine: bool,
    uv_taps: []const i16,
    uv_taps_q15: [constants.uv_max_taps]i32,
    ku: i32,
    kv: i32,
    level_black: i32,
    luma_numerator: i32,
    luma_denominator: i32,
    sine_q15: [2500]i16,

    pub fn init(standard: constants.Standard, setup: bool, precombine: bool) Error!Encoder {
        var result = Encoder{
            .standard = standard,
            .width = standard.activeWidth(),
            .denominator = standard.subcarrierDenominator(),
            .precombine = precombine,
            .uv_taps = undefined,
            .uv_taps_q15 = [_]i32{0} ** constants.uv_max_taps,
            .ku = 0,
            .kv = 0,
            .level_black = 0,
            .luma_numerator = 0,
            .luma_denominator = 0,
            .sine_q15 = [_]i16{0} ** 2500,
        };

        const level_white: i32 = switch (standard) {
            .pal => blk: {
                result.uv_taps = uv_taps_pal[0..];
                result.level_black = 0x4000;
                break :blk 0xD300;
            },
            .ntsc => blk: {
                result.uv_taps = uv_taps_ntsc[0..];
                result.level_black = if (setup) 0x4680 else 0x3C00;
                break :blk 0xC800;
            },
        };

        result.luma_numerator = @divTrunc(level_white - result.level_black, 256);
        result.luma_denominator = 219;
        if (result.luma_numerator * 256 != level_white - result.level_black) {
            result.luma_numerator = @divTrunc(level_white - result.level_black, 128);
            result.luma_denominator = 438;
        }

        const span = @as(f64, @floatFromInt(level_white - result.level_black));
        result.ku = @intFromFloat(@round((1.0 - 0.114) * kb / (112.0 * 256.0) * span * 32768.0));
        result.kv = @intFromFloat(@round((1.0 - 0.299) * kr / (112.0 * 256.0) * span * 32768.0));
        subcarrier.fillSineQ15(result.sine_q15[0..result.denominator]);
        for (result.uv_taps, 0..) |tap, i| result.uv_taps_q15[i] = tap;
        return result;
    }

    pub fn encodeLine(self: *const Encoder, destination: []u16, y: []const u16, u: []const u16, v: []const u16, phase: subcarrier.Line) void {
        std.debug.assert(destination.len >= self.width);
        std.debug.assert(y.len >= self.width and u.len >= self.width and v.len >= self.width);

        var u_staged: [constants.alignedWidth(928) + constants.uv_max_taps - 1]i32 = undefined;
        var v_staged: [constants.alignedWidth(928) + constants.uv_max_taps - 1]i32 = undefined;
        var u_filtered: [928]i32 = undefined;
        var v_filtered: [928]i32 = undefined;
        var luma: [928]i32 = undefined;
        const tap_count = self.uv_taps.len;
        const half = tap_count / 2;
        @memset(u_staged[0 .. constants.alignedWidth(928) + constants.uv_max_taps - 1], 0);
        @memset(v_staged[0 .. constants.alignedWidth(928) + constants.uv_max_taps - 1], 0);
        for (0..self.width) |x| {
            u_staged[x + half] = @as(i32, u[x]) - 32768;
            v_staged[x + half] = @as(i32, v[x]) - 32768;
        }
        fir.rowQ15(u_filtered[0..self.width], u_staged[0 .. self.width + tap_count - 1], self.uv_taps_q15[0..tap_count]);
        fir.rowQ15(v_filtered[0..self.width], v_staged[0 .. self.width + tap_count - 1], self.uv_taps_q15[0..tap_count]);

        const sine = @as(i32, self.sine_q15[phase.phase]);
        const cosine = @as(i32, self.sine_q15[(phase.phase + self.denominator / 4) % self.denominator]);
        const sine_pattern = [_]i32{ sine, cosine, -sine, -cosine };
        const cosine_pattern = [_]i32{ phase.v_switch * cosine, -phase.v_switch * sine, -phase.v_switch * cosine, phase.v_switch * sine };

        for (0..self.width) |x| {
            const luma_input = @as(i32, y[x]) - 4096;
            luma[x] = self.level_black + constants.roundDivSigned(@as(i64, luma_input) * self.luma_numerator, self.luma_denominator);
            const chroma_u = @divTrunc(u_filtered[x] * self.ku + 16384, 1 << 15);
            const chroma_v = @divTrunc(v_filtered[x] * self.kv + 16384, 1 << 15);
            const chroma = @divTrunc(chroma_u * sine_pattern[x & 3] + chroma_v * cosine_pattern[x & 3] + 16384, 1 << 15);
            destination[x] = constants.clampLevel(luma[x] + chroma);
        }
    }

    pub fn encodeFrame(self: *const Encoder, frame: i32, rows: usize, row_offset: i32, destination: []u16, destination_stride: usize, y: []const u16, y_stride: usize, u: []const u16, u_stride: usize, v: []const u16, v_stride: usize) void {
        std.debug.assert(destination.len >= destination_stride * rows);
        var filtered_u: [928]u16 = undefined;
        var filtered_v: [928]u16 = undefined;
        for (0..rows) |row| {
            const phase = subcarrier.line(self.standard, frame, row_offset + @as(i32, @intCast(row)));
            const dst = destination[row * destination_stride ..];
            const src_y = y[row * y_stride ..];
            if (!self.precombine) {
                self.encodeLine(dst, src_y, u[row * u_stride ..], v[row * v_stride ..], phase);
                continue;
            }

            const above = if (row >= 2) row - 2 else row;
            const below = if (row + 2 < rows) row + 2 else row;
            for (0..self.width) |x| {
                filtered_u[x] = @intCast((@as(u32, u[above * u_stride + x]) + 2 * @as(u32, u[row * u_stride + x]) + @as(u32, u[below * u_stride + x]) + 2) >> 2);
                filtered_v[x] = @intCast((@as(u32, v[above * v_stride + x]) + 2 * @as(u32, v[row * v_stride + x]) + @as(u32, v[below * v_stride + x]) + 2) >> 2);
            }
            self.encodeLine(dst, src_y, filtered_u[0..], filtered_v[0..], phase);
        }
    }
};

test "PAL encoder preserves neutral chroma around black" {
    var encoder = try Encoder.init(.pal, false, false);
    var y: [928]u16 = [_]u16{4096} ** 928;
    var u: [928]u16 = [_]u16{32768} ** 928;
    var v: [928]u16 = [_]u16{32768} ** 928;
    var output: [928]u16 = undefined;
    encoder.encodeLine(&output, &y, &u, &v, subcarrier.line(.pal, 0, 0));
    for (output) |sample| try std.testing.expectEqual(@as(u16, 0x4000), sample);
}

File: src/encode.rs
//! Fixed-point PAL and NTSC composite encoder.

use crate::model::{FrameIndex, GraySink, Plane, PlaneMut, Setup, Standard, YuvSource};
use crate::subcarrier::{line_phase, sine_table};

const MAX_WIDTH: usize = 928;
const MAX_TAPS: usize = 13;
const LEVEL_MIN: i32 = 0x0100;
const LEVEL_MAX: i32 = 0xfeff;

const PAL_UV_TAPS: [i32; 13] = [
    4, 38, 270, 1_226, 3_619, 6_927, 8_600, 6_927, 3_619, 1_226, 270, 38, 4,
];
const NTSC_UV_TAPS: [i32; 9] = [69, 626, 2_959, 7_563, 10_334, 7_563, 2_959, 626, 69];

/// A reusable, immutable composite encoder.
///
/// The encoder allocates only its phase table during construction. Frame and
/// row processing use fixed-size stack scratch whose upper bound is the PAL
/// active raster width.
#[derive(Debug)]
pub struct Encoder {
    standard: Standard,
    setup: Setup,
    precomb: bool,
    sine: Box<[i16]>,
    levels: LevelMap,
}

#[derive(Clone, Copy, Debug)]
struct LevelMap {
    black: i32,
    luma_num: i32,
    luma_den: i32,
    ku: i32,
    kv: i32,
}

impl Encoder {
    /// Constructs an encoder with a fixed standard and setup policy.
    #[must_use]
    pub fn new(standard: Standard, setup: Setup, precomb: bool) -> Self {
        let levels = LevelMap::new(standard, setup);
        let denominator = standard.raster().phase_denominator as usize;
        Self {
            standard,
            setup,
            precomb,
            sine: sine_table(denominator),
            levels,
        }
    }

    /// Returns the configured standard.
    #[must_use]
    pub const fn standard(&self) -> Standard {
        self.standard
    }

    /// Returns whether NTSC setup is enabled.
    #[must_use]
    pub const fn setup(&self) -> Setup {
        self.setup
    }

    /// Encodes a YUV444P16 active picture into a GRAY16 composite raster.
    ///
    /// All planes must have the standard's 4fsc width and the same height.
    /// NTSC permits either 480 active rows placed within the 486-line raster
    /// or the full 486 rows; `row_offset` carries that placement.
    pub fn encode(
        &self,
        frame: FrameIndex,
        row_offset: usize,
        y: Plane<'_>,
        u: Plane<'_>,
        v: Plane<'_>,
        mut output: PlaneMut<'_>,
    ) -> Result<(), EncodeError> {
        let dimensions = y.dimensions();
        if u.dimensions() != dimensions
            || v.dimensions() != dimensions
            || output.dimensions() != dimensions
        {
            return Err(EncodeError::PlaneGeometry {
                expected_width: self.standard.raster().width,
                y: dimensions,
                u: u.dimensions(),
                v: v.dimensions(),
                output: output.dimensions(),
            });
        }
        struct Source<'a> {
            y: Plane<'a>,
            u: Plane<'a>,
            v: Plane<'a>,
        }
        impl YuvSource for Source<'_> {
            fn dimensions(&self) -> (usize, usize) {
                self.y.dimensions()
            }
            fn row(&self, plane: usize, row: usize) -> &[u16] {
                match plane {
                    0 => self.y.row(row),
                    1 => self.u.row(row),
                    2 => self.v.row(row),
                    _ => panic!("invalid YUV plane"),
                }
            }
        }
        self.encode_source(frame, row_offset, &Source { y, u, v }, &mut output)
    }

    /// Encodes any row-addressable YUV444P16 source.
    pub fn encode_source<S: YuvSource, D: GraySink>(
        &self,
        frame: FrameIndex,
        row_offset: usize,
        source: &S,
        output: &mut D,
    ) -> Result<(), EncodeError> {
        let expected_width = self.standard.raster().width;
        let dimensions = source.dimensions();
        if dimensions.0 != expected_width || output.dimensions() != dimensions {
            return Err(EncodeError::PlaneGeometry {
                expected_width,
                y: dimensions,
                u: dimensions,
                v: dimensions,
                output: output.dimensions(),
            });
        }
        if row_offset + dimensions.1 > self.standard.raster().full_height {
            return Err(EncodeError::RasterPlacement {
                row_offset,
                rows: dimensions.1,
                raster_height: self.standard.raster().full_height,
            });
        }

        let mut filtered_u = [0_u16; MAX_WIDTH];
        let mut filtered_v = [0_u16; MAX_WIDTH];

        for row in 0..dimensions.1 {
            let (source_u, source_v) = if self.precomb {
                let above = row.checked_sub(2).unwrap_or(row);
                let below = (row + 2).min(dimensions.1 - 1);
                precomb_row(
                    source.row(1, above),
                    source.row(1, row),
                    source.row(1, below),
                    &mut filtered_u[..dimensions.0],
                );
                precomb_row(
                    source.row(2, above),
                    source.row(2, row),
                    source.row(2, below),
                    &mut filtered_v[..dimensions.0],
                );
                (&filtered_u[..dimensions.0], &filtered_v[..dimensions.0])
            } else {
                (source.row(1, row), source.row(2, row))
            };

            let mut encoded = [0_u16; MAX_WIDTH];
            self.encode_row(
                frame,
                row + row_offset,
                source.row(0, row),
                source_u,
                source_v,
                &mut encoded[..dimensions.0],
            );
            output.write_row(row, &encoded[..dimensions.0]);
        }
        Ok(())
    }

    fn encode_row(
        &self,
        frame: FrameIndex,
        raster_row: usize,
        y: &[u16],
        u: &[u16],
        v: &[u16],
        output: &mut [u16],
    ) {
        let width = output.len();
        let taps = match self.standard {
            Standard::Pal => &PAL_UV_TAPS[..],
            Standard::Ntsc => &NTSC_UV_TAPS[..],
        };
        let mut filtered_u = [0_i32; MAX_WIDTH];
        let mut filtered_v = [0_i32; MAX_WIDTH];
        fir_chroma_row(u, taps, &mut filtered_u[..width]);
        fir_chroma_row(v, taps, &mut filtered_v[..width]);

        let carrier = line_phase(self.standard, frame, raster_row);
        let den = self.standard.raster().phase_denominator as usize;
        let sin = i32::from(self.sine[carrier.phase]);
        let cos = i32::from(self.sine[(carrier.phase + den / 4) % den]);
        let sine4 = [sin, cos, -sin, -cos];
        let cosine4 = [
            carrier.v_switch * cos,
            -carrier.v_switch * sin,
            -carrier.v_switch * cos,
            carrier.v_switch * sin,
        ];

        for (x, destination) in output.iter_mut().enumerate() {
            let luma = self.levels.black
                + round_div(
                    i64::from(i32::from(y[x]) - 4_096) * i64::from(self.levels.luma_num),
                    self.levels.luma_den,
                );
            let chroma_u = (filtered_u[x] * self.levels.ku + 16_384) >> 15;
            let chroma_v = (filtered_v[x] * self.levels.kv + 16_384) >> 15;
            let chroma = (chroma_u * sine4[x & 3] + chroma_v * cosine4[x & 3] + 16_384) >> 15;
            *destination = (luma + chroma).clamp(LEVEL_MIN, LEVEL_MAX) as u16;
        }
    }
}

impl LevelMap {
    fn new(standard: Standard, setup: Setup) -> Self {
        let (black, white) = match standard {
            Standard::Pal => (0x4000, 0xd300),
            Standard::Ntsc => (
                if setup == Setup::Ire7_5 {
                    0x4680
                } else {
                    0x3c00
                },
                0xc800,
            ),
        };
        let mut luma_num = (white - black) / 256;
        let mut luma_den = 219;
        if luma_num * 256 != white - black {
            luma_num = (white - black) / 128;
            luma_den = 438;
        }
        let span = f64::from(white - black);
        let ku = (((1.0 - 0.114) * 0.492_111_041_122_483_56 / (112.0 * 256.0)) * span * 32_768.0)
            .round() as i32;
        let kv = (((1.0 - 0.299) * 0.877_283_219_938_178_7 / (112.0 * 256.0)) * span * 32_768.0)
            .round() as i32;
        Self {
            black,
            luma_num,
            luma_den,
            ku,
            kv,
        }
    }
}

fn precomb_row(above: &[u16], current: &[u16], below: &[u16], output: &mut [u16]) {
    for (((destination, &a), &c), &b) in output.iter_mut().zip(above).zip(current).zip(below) {
        *destination = ((u32::from(a) + 2 * u32::from(c) + u32::from(b) + 2) >> 2) as u16;
    }
}

fn fir_chroma_row(input: &[u16], taps: &[i32], output: &mut [i32]) {
    debug_assert_eq!(input.len(), output.len());
    debug_assert!(taps.len() <= MAX_TAPS);
    let radius = taps.len() / 2;
    let length = input.len();
    let interior_start = radius.min(length);
    let interior_end = length.saturating_sub(radius);

    if interior_start >= interior_end {
        for x in 0..length {
            output[x] = fir_chroma_edge(input, taps, radius, x);
        }
        return;
    }

    for x in 0..interior_start {
        output[x] = fir_chroma_edge(input, taps, radius, x);
    }
    for x in interior_start..interior_end {
        output[x] = fir_chroma_interior(input, taps, radius, x);
    }
    for x in interior_end..length {
        output[x] = fir_chroma_edge(input, taps, radius, x);
    }
}

fn fir_chroma_edge(input: &[u16], taps: &[i32], radius: usize, x: usize) -> i32 {
    let first_tap = radius.saturating_sub(x);
    let last_tap = (input.len() + radius - x).min(taps.len());
    let mut accumulator = 0_i64;
    for (tap_index, &tap) in taps[first_tap..last_tap].iter().enumerate() {
        let sample = input[x + first_tap + tap_index - radius];
        accumulator += i64::from(tap) * i64::from(i32::from(sample) - 32_768);
    }
    ((accumulator + 16_384) >> 15) as i32
}

fn fir_chroma_interior(input: &[u16], taps: &[i32], radius: usize, x: usize) -> i32 {
    let samples = &input[x - radius..x + radius + 1];
    let (tap_chunks, tap_tail) = taps.as_chunks::<4>();
    let (sample_chunks, sample_tail) = samples.as_chunks::<4>();
    let mut accumulator = 0_i64;
    for (tap_chunk, sample_chunk) in tap_chunks.iter().zip(sample_chunks) {
        accumulator += i64::from(tap_chunk[0])
            * i64::from(i32::from(sample_chunk[0]) - 32_768);
        accumulator += i64::from(tap_chunk[1])
            * i64::from(i32::from(sample_chunk[1]) - 32_768);
        accumulator += i64::from(tap_chunk[2])
            * i64::from(i32::from(sample_chunk[2]) - 32_768);
        accumulator += i64::from(tap_chunk[3])
            * i64::from(i32::from(sample_chunk[3]) - 32_768);
    }
    for (&tap, &sample) in tap_tail.iter().zip(sample_tail) {
        accumulator += i64::from(tap) * i64::from(i32::from(sample) - 32_768);
    }
    ((accumulator + 16_384) >> 15) as i32
}

pub(crate) fn round_div(numerator: i64, denominator: i32) -> i32 {
    let half = i64::from(denominator / 2);
    ((if numerator >= 0 {
        numerator + half
    } else {
        numerator - half
    }) / i64::from(denominator)) as i32
}

pub(crate) fn level_parameters(standard: Standard, setup: Setup) -> (i32, i32, i32, i32, i32) {
    let encode = LevelMap::new(standard, setup);
    let decode_luma_num = encode.luma_den;
    let decode_luma_den = encode.luma_num;
    (
        encode.black,
        decode_luma_num,
        decode_luma_den,
        encode.ku,
        encode.kv,
    )
}

/// Invalid encoder plane geometry or raster placement.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum EncodeError {
    /// The planes disagree or do not use the standard's active width.
    PlaneGeometry {
        /// Required composite raster width for the selected standard.
        expected_width: usize,
        /// Actual luma-plane \`(width, height)\`.
        y: (usize, usize),
        /// Actual U-plane \`(width, height)\`.
        u: (usize, usize),
        /// Actual V-plane \`(width, height)\`.
        v: (usize, usize),
        /// Actual output-plane \`(width, height)\`.
        output: (usize, usize),
    },
    /// Active rows extend beyond the broadcast raster.
    RasterPlacement {
        /// First active row within the complete raster.
        row_offset: usize,
        /// Number of active rows in the input frame.
        rows: usize,
        /// Complete active-picture height for the selected standard.
        raster_height: usize,
    },
}

impl std::fmt::Display for EncodeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::PlaneGeometry {
                expected_width,
                y,
                u,
                v,
                output,
            } => write!(
                f,
                "encoder needs matching {expected_width}-sample planes; Y={y:?}, U={u:?}, V={v:?}, output={output:?}"
            ),
            Self::RasterPlacement {
                row_offset,
                rows,
                raster_height,
            } => write!(
                f,
                "rows {row_offset}..{} exceed the {raster_height}-line raster",
                row_offset + rows
            ),
        }
    }
}

impl std::error::Error for EncodeError {}

#[cfg(test)]
mod tests {
    use super::*;

    fn fir_chroma_row_reference(input: &[u16], taps: &[i32], output: &mut [i32]) {
        for (x, destination) in output.iter_mut().enumerate() {
            let mut accumulator = 0_i64;
            for (tap_index, &tap) in taps.iter().enumerate() {
                let source = x + tap_index;
                if source < taps.len() / 2 || source - taps.len() / 2 >= input.len() {
                    continue;
                }
                accumulator += i64::from(tap)
                    * i64::from(i32::from(input[source - taps.len() / 2]) - 32_768);
            }
            *destination = ((accumulator + 16_384) >> 15) as i32;
        }
    }

    #[test]
    fn chroma_filters_preserve_a_flat_signal() {
        for taps in [&PAL_UV_TAPS[..], &NTSC_UV_TAPS[..]] {
            let input = [40_000_u16; 64];
            let mut output = [0_i32; 64];
            fir_chroma_row(&input, taps, &mut output);
            for value in &output[taps.len()..64 - taps.len()] {
                assert_eq!(*value, 7_232);
            }
        }
    }

    #[test]
    fn split_chroma_filter_matches_reference_at_all_boundaries() {
        for taps in [&PAL_UV_TAPS[..], &NTSC_UV_TAPS[..]] {
            for length in 0..=40 {
                let input: Vec<_> = (0..length)
                    .map(|index| (index * 4_123 + 7_000) as u16)
                    .collect();
                let mut expected = vec![0_i32; length];
                let mut actual = vec![0_i32; length];
                fir_chroma_row_reference(&input, taps, &mut expected);
                fir_chroma_row(&input, taps, &mut actual);
                assert_eq!(actual, expected, "length={length}, taps={}", taps.len());
            }
        }
    }

    #[test]
    fn round_division_is_symmetric() {
        for value in 0..1000 {
            assert_eq!(round_div(value, 219), -round_div(-value, 219));
        }
    }
}
